[Madur] iP

CraveToCode opened the pull request:
### Duchess
> "It does not do to dwell on dreams and forget to live." - Albus Dumbledore

Duchess frees your mind from having to dwell on things you need to do and allows you to live. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from here [(source)](https://github.com/CraveToCode/ip/releases)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Duke(FILE_PATH).run();
}
```

Commit:
Add Gradle support

CraveToCode committed:
Level-1

CraveToCode committed:
Added Task.java, addTask, printTaskList.

CraveToCode committed:
Added feature to mark tasks as done.

CraveToCode committed:
Added ToDo.java, Deadline.java, Event.java.
Added ability to add todos, deadlines, events.

CraveToCode committed:
Added some Automated Text UI Testing.

CraveToCode committed:
Added DukeException class. Added exception handling for missing inputs / arrayoutofbounderror.

CraveToCode committed:
Added Delete functionality. Updated Automatic Text UI Testing.

CraveToCode committed:
Added pretty printing of output. Added missingInput in DukeException.

CraveToCode committed:
Move Duke logo location

CraveToCode committed:
Add save functionality of tasks to hard disk

CraveToCode committed:
Change storage of date and time to LocalDate objects

CraveToCode committed:
Merge branch 'branch-Level-7'

CraveToCode committed:
Merge branch 'branch-Level-8' to 'master'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Todo.java

CraveToCode committed:
Fix merge conflict bugs from merging branch-Level-8

CraveToCode committed:
Add more OOP for parsing, ui, storage, commands

CraveToCode committed:
Change ui showErrorMessage to static method

CraveToCode committed:
Update packages of all files relevant to duke

CraveToCode committed:
Add JUnit tests for Storage and Datetime

CraveToCode committed:
Package app as JAR file

CraveToCode committed:
Add find command as an available command option

CraveToCode committed:
Add JavaDocs for all classes

CraveToCode committed:
Merge branch 'branch-A-JavaDoc'

CraveToCode committed:
Merge 'branch-Level-9' and resolve merge conflicts

CraveToCode committed:
Add JavaDocs for FindCommand

CraveToCode committed:
Merge branch 'add-gradle-support'

CraveToCode committed:
Update build.gradle mainClassName and remove checkstyle config

CraveToCode committed:
Merge branch 'branch-A-Gradle'

CraveToCode committed:
Add checkstyle config files

CraveToCode committed:
Fix code style issues

CraveToCode committed:
Merge branch 'branch-A-Checkstyle'

twothicc commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,47 @@
+import duke.functionality.Storage;
+import duke.functionality.Ui;
+import duke.functionality.Parser;
+import duke.commands.*;
```
Perhaps it would be better to list imported classes explicitly here.

This could help users in comprehending and maintaining the class as the project grows.

twothicc commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+import duke.exceptions.DukeException;
+import duke.functionality.Ui;
+import duke.functionality.Storage;
+import duke.tasks.*;
```
Perhaps it would be better to list imported classes explicitly here.

This could help users in comprehending and maintaining the class as the project grows.

twothicc commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public void execute(Storage storage, Ui ui) {
+        if (this.taskType.equals(Todo.taskTag())) {
+            int taskListLen = storage.taskListLen();
```
I noticed that you are calling storage.taskListLen method at the start of each of these code blocks.
Perhaps it would be better to just call it once at the start to shorten code length and improve readability.

twothicc commented on src/main/java/duke/functionality/Parser.java:
```diff
@@ -0,0 +1,98 @@
+        return inputSplit.length < 2;
+    }
+
+    private static void timeFormatChecker(String inputString) throws DukeException {
```
Perhaps a verb rather than a noun might be a more appropriate name for this method.

twothicc commented on src/main/java/duke/functionality/Storage.java:
```diff
@@ -0,0 +1,123 @@
+public class Storage {
+    private final static String FILE_NAME = "dukestorage.txt";
+    private String filePath;
+    private File f;
```
Perhaps file would be a more descriptive name.

twothicc commented on src/main/java/duke/functionality/Ui.java:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * Sends a welcome message to the user.
+     */
+    public void welcomeMessage() {
```
Perhaps a verb rather than a noun would be a more suitable name for this method which prints your standard welcome message to the user.

This applies to other user interaction related methods in this class.

twothicc commented on src/main/java/duke/functionality/Ui.java:
```diff
@@ -0,0 +1,158 @@
+ * Creates an Ui system that is responsible for interacting with the user.
+ */
+public class Ui {
+    private static final String BORDER = "____________________________________________________________";
+    private static final String NEW_LINE = System.lineSeparator();
+    private static final String PLACEHOLDER = "%text%";
+    private static final String DONE_TASK = "Nice! I've marked this task as done:";
+    private static final String MAX_TASK = "Sorry! You have max number of tasks stored already.";
+    private static final String ADDED_TASK = "Got it. I've added this task:";
+    private static final String REMAINING_TASK_NUM = String.format("Now you have %s tasks in the list.", PLACEHOLDER);
+    private static final String DELETED_TASK = "Noted. I've removed this task:";
+    private static final String MISSING_TASK = "There is no task at the specified index.";
+    private static final String NO_MATCHING_TASKS = "There are no matching tasks found.";
+    private static final String UNKNOWN_COMMAND = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
```
Perhaps it would be better to prefix certain names to constants that are related to exception reporting, user interaction and formatting.

This could help you maintain a large number of constants as your project scales in size.

twothicc commented on src/main/java/duke/functionality/Parser.java:
```diff
@@ -0,0 +1,98 @@
+package duke.functionality;
+
+import duke.commands.*;
```
Perhaps it would be better to import classes explicitly.

twothicc commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+            int taskListLen = storage.taskListLen();
+            if (taskListLen < TaskList.MAX_TASKS) {
+                Todo newTodo = new Todo(this.taskName);
+                storage.addTask(newTodo);
+                taskListLen += 1;
+                storage.saveToFile();
+                ui.taskAddedMessage(newTodo, taskListLen);
+            } else {
+                ui.maxTaskReachedMessage();
+            }
```
I noticed that you have a lot of repeating code here.
Perhaps it would be better to handle creating the correct task first, then handle to repeated logic.

This would improve code readability.

kaushikkrdy replied:
Perhaps you extract this repeated portion into another function that you can call each time to reduce code repetition?

twothicc reviewed (commented):
LGTM :thumbsup:
Only some recommended changes to improve code readability and standard.

kaushikkrdy commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+        super(message);
+    }
+
+    public static DukeException missingInput(String taskName) {
```
Should this method have a header comment since it is a public method? (in line with Java coding standards)

kaushikkrdy commented on src/main/java/duke/commands/UnknownCommand.java:
```diff
@@ -0,0 +1,26 @@
+        ui.unknownCommandMessage();
+    }
+
+    public boolean isExit() {
```
Perhaps isExit() could have a header comment in all Command classes since it is a public method?

kaushikkrdy reviewed (commented):
LGTM! The code is well-organised apart from a few minor coding standard issues.

CraveToCode committed:
Add GUI implementation and refactor code accordingly

CraveToCode committed:
Fix codestyle issues

CraveToCode committed:
Add more JavaDocs and fix more codestyle issues

CraveToCode committed:
Merge branch 'branch-Level-10'

CraveToCode committed:
Rename function to fit code style standards

CraveToCode committed:
Fix minor format issues

CraveToCode committed:
Refactor code to improve code quality

The filtering of user input command keywords and the subsequent actions
taken to create a command is being done through one entire if-else
block.

This is difficult to expand upon in a manageable manner in the future
when there might be more commands available.

Filtering of user input keywords is being achieved through an if-else
block, but the subsequent actions taken after the filtering process
is being extracted into separate commands.

This will allow future command preparations to be easily added into
the parser in the form of a separate method.

CraveToCode committed:
Merge pull request #1 from CraveToCode/branch-A-Assertions

Add assertions to code

CraveToCode committed:
Merge branches 'master' and 'master' of https://github.com/CraveToCode/ip

CraveToCode committed:
Merge branch 'master' into branch-A-CodeQuality

CraveToCode committed:
Merge pull request #2 from CraveToCode/branch-A-CodeQuality

Refactor code to improve code quality

CraveToCode committed:
Add Github Continuous Integration (CI)

CraveToCode committed:
Add capability to archive tasks

System does not allow users to archive tasks.

Creating a new file to save tasks that user wants to archive. Adding
necessary read/write/display commands to allow users to archive tasks,
list archived tasks, and restore archived tasks.

CraveToCode committed:
Complete C Extension to allow users to archive tasks

CraveToCode committed:
Merge branch 'branch-BCD-Extension'

CraveToCode committed:
Update GUI with better visuals

CraveToCode committed:
Add Ui image and title for Main Window of GUI

CraveToCode committed:
Update README.md and add relevant GUI images

CraveToCode committed:
Fix filepath for Ui.png

CraveToCode committed:
Edit README.md

CraveToCode committed:
Set theme jekyll-theme-cayman

CraveToCode committed:
Delete _config.yml

CraveToCode committed:
Merge branch 'master' of https://github.com/CraveToCode/ip

CraveToCode committed:
Update README.md

CraveToCode committed:
Update README.md

CraveToCode committed:
Update README.md

CraveToCode committed:
Merge branch 'master' of https://github.com/CraveToCode/ip

CraveToCode committed:
Fix index bug in archive and restore commands

CraveToCode committed:
Fix index bug in restore and archive commands

CraveToCode committed:
Update README.md
